ext/standard/string.c

```diff
@@ -1403,13 +1406,95 @@ PHPAPI char *php_strtoupper(char *s, size_t len)
 }
 /* }}} */
 
+#ifdef __aarch64__
+
+static zend_always_inline uint8x16_t neon_to_lu(uint8x16_t x,
+						uint8_t shift_LUT[4][64],
+						uint8_t has_shift[4]) {
+	uint8x16_t shifts = {0};
+
+	/* There are no lower/upper case chars in the first table: [0, 63]. */
+	int i = 1;
+	for (; i < 4; i++) {
+		if (!has_shift[i])
+			continue;
+
+		/* Mask out the higher bits that are used to index in the
+		   shift_LUT[i]: a vector table lookup that accesses in the
+		   range [0..63] will return the value recorded in the LUT,
+		   otherwise the returned value is 0. The value in the LUT is
+		   the shift amount to convert low->up or up->low case. */
+		uint8x16_t idx = veorq_u8(x, vdupq_n_u8(i << 6));
+		uint8x16_t s = vqtbl4q_u8(*((uint8x16x4_t *)shift_LUT[i]), idx);
+		shifts = vaddq_u8(shifts, s);
+	}
+	return vaddq_u8(x, shifts);
+}
+
+static zend_always_inline
+zend_string *neon_string_to_lu(zend_string *s,
+			       uint8_t shift_LUT[4][64],
+			       uint8_t has_shift[4])
+{
+	size_t len = ZSTR_LEN(s);
+	zend_string *res = zend_string_alloc(len, 0);
+	uint8_t *in = (uint8_t *) ZSTR_VAL(s);
+	uint8_t *out = (uint8_t *) ZSTR_VAL(res);
+	int j = 0;
+
+	for (; len - j > 15; j += 16)
+		vst1q_u8((uint8_t *)(out + j),
+			 neon_to_lu(*((uint8x16_t *)(in + j)),
+				    shift_LUT, has_shift));
+
+	/* Process 15 or fewer bytes. */
+	if (j < len) {
+		uint8_t left_bytes = len - j;
+		uint8x16_t x;
+
+		ZEND_ASSERT(left_bytes < 16);
+		memcpy((uint8_t *)&x, in + j, left_bytes);
+		x = neon_to_lu(x, shift_LUT, has_shift);
+		memcpy(out + j, (uint8_t *)&x, left_bytes);
+	}
+
+	return res;
+}
+#endif /* __aarch64__ */
+
+
 /* {{{ php_string_toupper
  */
 PHPAPI zend_string *php_string_toupper(zend_string *s)
 {
 	unsigned char *c;
 	const unsigned char *e;
 
+#ifdef __aarch64__
+	static uint8_t toupper_shift_LUT[4][64] = {0};
+	static uint8_t toupper_has_shift[4] = {0};
+	static char *saved_locale = NULL;
```

Comment:
Caching in static variables, won't work for thread-safe (ZTS) build.

Reply:
Thanks @dstogov for catching that.
I added ifndef ZTS.  Ok to commit?

Reply:
I'm not sure.
The original implementation avoids string duplication, if it's already in lower case.
So, this patch may make negative effect.

Reply:
Fixed in the last patch.